### PR TITLE
Replace error callback with Result

### PR DESCRIPTION
**Pull request by Mark-Simulacrum**
r? @petrochenkov

### CODE REVIEW

File: src/librustc_resolve/lib.rs [hunk 1]
```diff
@@ -1855,48 +1858,46 @@ impl<'a> Resolver<'a> {
                     .collect(),
             }
         };
-        let res = self.resolve_ast_path_cb(&path, is_value, |_, _, _| errored = true);
-        if errored || res == def::Res::Err {
-            Err(())
-        } else {
-            Ok((path, res))
+        match self.resolve_ast_path_inner(&path, is_value) {
+            Ok(res) => {
```
Comment (petrochenkov) [inline src/librustc_resolve/lib.rs]:
```suggestion
            Ok(res) if res != Res::Err => {
```
Although, I'd rather remove this condition (and then use `?`).
If the function returns `Res`, then rustdoc should be ready to deal with `Res::Err` as well.

**Comment by Mark-Simulacrum** [comment]
Removed the condition, mostly by uplifting it into rustdoc. I'm not sure if those changes are a correct handling of this -- the code looks a bit questionable to me -- but I don't think this is quite the place to refactor rustdoc's resolution too much :) Happy to do so if you disagree, though!

File: src/librustc_resolve/lib.rs [hunk 2]
```diff
@@ -1855,48 +1858,38 @@ impl<'a> Resolver<'a> {
...
-            Ok((path, res))
-        }
+        let res = self.resolve_ast_path_inner(&path, is_value).map_err(|_| ())?;
+        Ok((path, res))
     }
 
     /// Like `resolve_ast_path`, but takes a callback in case there was an error.
-    // FIXME(eddyb) use `Result` or something instead of callbacks.
-    fn resolve_ast_path_cb<F>(
+    fn resolve_ast_path_inner(
```
Comment (eddyb) [inline src/librustc_resolve/lib.rs]: The comment wasn't updated. But other than that I'm really happy the callback is gone!

Comment (Mark-Simulacrum) [reply src/librustc_resolve/lib.rs]: I think it was lost in a rebase. I've added to my todo list to check if it's been removed after some of the resolver refactorings land so I don't cause merge conflicts and such.

**Pull request merged**